src/audio/mod.rs

```diff
@@ -1,44 +1,41 @@
-use crate::{Pipeline, Result};
+use crate::{ring_buffer, Pipeline, Result};
 use anyhow::anyhow;
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
-    Device,
+    Device, StreamConfig,
 };
-use ringbuf::{Consumer, Producer, RingBuffer};
+use ringbuf::RingBuffer;
 use std::time::Duration;
 
 const LATENCY: f32 = 20.0;
 
-pub fn run(mut pipeline: Pipeline) -> Result<()> {
-    let (input_device, output_device) = devices()?;
-
-    let config: cpal::StreamConfig = input_device.default_input_config()?.into();
-
+pub fn run(
+    input_device: &Device,
+    output_device: &Device,
+    config: &StreamConfig,
+    mut pipeline: Pipeline,
+) -> Result<()> {
     let latency_num_frames = (LATENCY / 1_000.0) * config.sample_rate.0 as f32;
     let latency_num_samples = latency_num_frames as usize * config.channels as usize;
 
     let ring = RingBuffer::new(latency_num_samples * 2);
     let (mut producer, mut consumer) = ring.split();
 
-    for _ in 0..latency_num_samples {
-        producer.push(0.0).unwrap();
-    }
+    ring_buffer::write_empty_samples(&mut producer, latency_num_samples)?;
```

Comment:
I'm using a ring buffer in the delay pedal too, so I added some convenience functions.